src/client.rs

```diff
@@ -685,6 +687,69 @@ pub trait UiClient: PollClient {
     }
 }
 
+pub struct ClientBuilder<B: 'static = ()> {
+    id: PathBuf,
+    backends: &'static [BackendId<B>],
+}
+
+impl ClientBuilder {
+    pub fn new(id: impl Into<PathBuf>) -> Self {
+        Self {
+            id: id.into(),
+            backends: &[],
+        }
+    }
+}
+
+impl<B: 'static> ClientBuilder<B> {
+    pub fn backends<C: 'static>(self, backends: &'static [BackendId<C>]) -> ClientBuilder<C> {
+        ClientBuilder {
+            id: self.id,
+            backends,
+        }
+    }
+
+    #[allow(clippy::result_unit_err)]
+    fn create_endpoint<P: Platform, D: Dispatch<P, BackendId = B>>(
+        self,
+        service: &mut Service<P, D>,
+    ) -> Result<Requester<TrussedInterchange>, ()> {
+        let (requester, responder) = TrussedInterchange::claim().ok_or(())?;
+        service
+            .add_endpoint(responder, self.id, self.backends)
+            .map_err(|_service_endpoint| ())?;
+        Ok(requester)
+    }
+
+    #[allow(clippy::result_unit_err)]
+    pub fn build<P: Platform, D: Dispatch<P, BackendId = B>, S: Syscall>(
+        self,
+        service: &mut Service<P, D>,
+        syscall: S,
+    ) -> Result<ClientImplementation<S>, ()> {
```

Comment:
In these Results with trivial error types, maybe we should use Option and mark with `#[must_use]`?
I'm finding the proliferation of `()` a bit hard to follow.
Or would this break `?` in certain places?

Reply:
I just used `()` for consistency with the existing methods, but I agree that `Error` is much better.  I’ll add a `ClientCountExceeded` variant and change all relevant methods in `ClientBuilder` and `Service` to return that error instead.

Reply:
Yeah that works. Generally it'd be better to have tight error types (i.e. a client of just this method would know which error variants it can actually expect), but doing this in Rust is tedious as you'd need tons of conversion traits.